wallet-new/integration/Functions.hs

```diff
@@ -21,467 +25,556 @@ import           Cardano.Wallet.API.V1.Types (Account (..), AccountIndex, Accoun
+                                        getTransactionIndex, getWallets, hoistClient)
 
-import           Pos.Core (mkCoin)
-import           Pos.Util (maybeThrow)
+import           Pos.Core (getCoin, mkCoin)
 import qualified Pos.Wallet.Web.ClientTypes.Types as V0
 
 import           Error
 import           Types
 
-
--- | The top function that we need to run in order
--- to test the backend.
+newtype RefT s m a
+    = RefT
+    { unRefT :: ReaderT (IORef s) m a
+    } deriving
+    ( Functor, Applicative, Monad, MonadIO, MonadTrans, Alternative, MonadCatch
+    , MonadThrow, MonadPlus
+    )
+
+instance MonadIO m => MonadState s (RefT s m) where
+    get = RefT ask >>= readIORef
+    put x = RefT ask >>= (`writeIORef` x)
+
+runRefT :: MonadIO m => RefT s m a -> s -> m (a, s)
+runRefT act i = do
+    init <- newIORef i
+    a <- runReaderT (unRefT act) init
+    s <- readIORef init
+    pure (a, s)
+
+execRefT :: MonadIO m => RefT s m a -> s -> m s
+execRefT x = fmap snd . runRefT x
+
+-- | This function uses the given 'ActionProbabilities' to generate a list
+-- of actions to test against the API. It runs the actions through,
+-- eventually returning the final state.
 runActionCheck
-    :: (WalletTestMode m)
+    :: (WalletTestMode m, HasCallStack)
     => WalletClient m
     -> WalletState
     -> ActionProbabilities
     -> m WalletState
 runActionCheck walletClient walletState actionProb = do
-    action <- chooseAction actionProb
-    runAction walletClient walletState action
-
+    actions <- chooseActions 100 actionProb
+    let client' = hoistClient lift walletClient
+    execRefT (tryAll (map (runAction client') actions)) walletState
+        `catch` \x -> fmap (const walletState) . liftIO . hspec .
+            describe "Unit Test Failure" $
+                it ("threw a test error: " ++ showConstr x) $ case x of
+                    LocalWalletDiffers a b ->
+                        a `shouldBe` b
+                    LocalWalletsDiffers as bs ->
+                        sort as `shouldBe` sort bs
+                    LocalAccountDiffers a b ->
+                        a `shouldBe` b
+                    LocalAccountsDiffers as bs ->
+                        sort as `shouldBe` sort bs
+                    LocalAddressesDiffer as bs ->
+                        sort as `shouldBe` sort bs
+                    LocalAddressDiffer a b ->
+                        a `shouldBe` b
+                    LocalTransactionsDiffer as bs ->
+                        sort as `shouldBe` sort bs
+                    LocalTransactionMissing txn txns ->
+                        txns `shouldContain` [txn]
+                    err ->
+                        expectationFailure $ show err
+
+-- | Attempt each action in the list. If an action fails, ignore the
+-- failure and try the next action in the sequence.
+tryAll :: Alternative f => NonEmpty (f a) -> f a
+tryAll (x :| xs) = foldl' (\this next -> (this *> next) <|> next) x xs
```

Comment:
This implementation of `tryAll` does not have the problems of the old one, and the `NonEmpty` constraint makes it even safer!